### ISSUE TITLE
Nullpointerexception for mapped optional value case classes in right join

### ISSUE THREAD

**Issue by valenterry** [tpolecat/doobie#551]
When doing a right join and mapping the result into the specified case class, doobie seems to construct the fields of this case classes with `null`-values - to replace them with `None` after that. However, just initializing a case class can result in side-effects which can be a problem. I could minify it to the following example, using
```
"org.tpolecat" %% "doobie-core-cats" % "0.4.1",
"org.tpolecat" %% "doobie-postgres-cats" % "0.4.1",
```

Here is my code:
```
    case class Foo(value: String)
    case class Bar(value: String) {
       assert(value != null) //We could use any kind of constraints here. Removing this makes the test succeed.
    }
    case class FooBarRow(foo: Foo, bar: Option[Bar])

    val transactor = DriverManagerTransactor[IOLite](???)

    val result = sql"""
      SELECT
        table_foo.foo,
        table_bar.bar
      FROM table_bar
      RIGHT JOIN (
        SELECT
          foo
        FROM table_foo
      ) table_foo
      ON table_bar.foo = table_foo.foo
    """.query[FooBarRow]

    result
      .vector
      .transact(transactor)
      .attempt
      .unsafePerformIO match {
        case Right(foobarrows) => //Works if we remove the assert in Bar case class
          assert( foobarrows == Vector(FooBarRow(Foo("doobiefoo"), None)) )
        case Left(throwable) => //This is called when the right join leads to a row where table_bar.bar is null
          ???
      }
```

I used the following SQL to create the tables for executing the above SQL:
```
CREATE TABLE table_foo (
    foo text,
    PRIMARY KEY(foo)
);

CREATE TABLE table_bar (
    bar text,
    foo text unique,
    PRIMARY KEY(bar)
);

insert into table_foo (foo) values ('doobiefoo');
```

So it seems that doobie calls `Bar(null)` but it discards it and simply returns `None` so the endresult is `FooBarRow(Foo("doobiefoo"), None)` as expected. Unless `Bar(null)` blows up with an exception because of constraints or with a NPE when the value is somewhere used in the constructor.

**Comment by wedens** [comment]
related https://github.com/tpolecat/doobie/pull/540

**Comment by tpolecat** [comment]
Yes, this is a known issue in 0.4.1 and earlier. 

```scala
// 0.4.1
scala>  case class Bar(value: String) {
     |        assert(value != null) //We could use any kind of constraints here. Removing this makes the test succeed.
     |     }
defined class Bar

scala> sql"select null".query[Option[Bar]].unique.quick.unsafePerformIO
java.lang.AssertionError: assertion failed
  at scala.Predef$.assert(Predef.scala:204)
  ... 68 elided
 
scala> sql"select 'blah'".query[Option[Bar]].unique.quick.unsafePerformIO
  Some(Bar(blah))
```

The workaround is to create a pair of `Atom` instances for your data type rather than using the default `Meta` mapping. 

This is fixed in 0.4.2.

```scala
// 0.4.2-SNAPSHOT
scala>  case class Bar(value: String) {
     |        assert(value != null) //We could use any kind of constraints here. Removing this makes the test succeed.
     |     }
defined class Bar
 
scala> sql"select null".query[Option[Bar]].unique.quick.unsafePerformIO
  None
 
scala> sql"select 'blah'".query[Option[Bar]].unique.quick.unsafePerformIO
  Some(Bar(blah))
```
